In [1]:
import numpy as np
import pandas as pd
from numpy.linalg import inv
from numpy import dot

In [2]:
df = pd.read_csv('Combined.csv') 

In [3]:
df.head()

observation_date  DHHNGSP  Mkt-RF  OVXCLS     Yt
0         01/03/17     3.41    0.83   32.88  55.05
1         01/04/17     3.42    0.79   32.36  54.57
2         01/05/17     3.42   -0.21   31.56  54.99
3         01/06/17     3.38    0.29   30.47  55.90
4         01/09/17     3.14   -0.37   33.73  54.39

In [4]:
temp = df.iloc[:,1:5]
temp['x0'] = 1
temp.head()
temp

DHHNGSP  Mkt-RF  OVXCLS     Yt  x0
0       3.41    0.83   32.88  55.05   1
1       3.42    0.79   32.36  54.57   1
2       3.42   -0.21   31.56  54.99   1
3       3.38    0.29   30.47  55.90   1
4       3.14   -0.37   33.73  54.39   1
5       3.21    0.16   33.61  53.20   1
6       3.27    0.31   32.95  53.61   1
7       3.36   -0.30   32.37  54.51   1
8       3.36    0.29   30.73  54.37   1
9       3.36   -0.49   31.25  54.30   1
10      3.37    0.24   31.24  54.68   1
11      3.26   -0.38   30.88  53.77   1
12      3.25    0.33   29.19  53.18   1
13      3.23   -0.29   30.79  55.04   1
14      3.16    0.83   29.59  54.80   1
15      3.16    0.84   30.41  54.70   1
16      3.26   -0.10   29.59  54.34   1
17      3.44   -0.12   29.78  55.89   1
18      3.31   -0.68   30.94  54.80   1
19      3.22    0.00   31.33  54.77   1
20      3.00    0.04   30.65  55.25   1
21      3.15   -0.02   29.39  55.79   1
22      3.13    0.82   28.11  55.94   1
23      3.13   -0.27   29.37  55.92   1
24      2.92   -0.01   29.61  55.02   1
25      3.03    0.06   29.09  53.79   1
26      3.05    0.72   28.25  54.42   1
27      3.11    0.38   26.43  53.98   1
28      3.11    0.49   26.99  55.20   1
29      2.93    0.45   27.49  54.15   1
..       ...     ...     ...    ...  ..
221     3.16    1.01   28.42  64.49   1
222     3.18   -0.14   26.96  64.35   1
223     3.17    0.21   27.18  62.94   1
224     3.13    0.67   26.64  60.91   1
225     3.06   -0.05   26.83  61.25   1
226     3.16    0.21   26.88  61.18   1
227     3.09   -0.06   28.36  61.34   1
228     3.04    1.06   28.13  61.34   1
229     3.04    0.02   28.02  62.28   1
230     3.05    0.82   25.95  62.92   1
231     3.05   -0.22   23.73  63.27   1
232     3.05   -0.10   23.68  63.58   1
233     3.05   -0.43   23.13  63.25   1
234     2.88   -0.09   23.13  63.56   1
235     2.88    0.42   23.13  63.74   1
236     3.06    0.51   23.85  63.53   1
237     3.06    0.26   22.96  64.57   1
238     2.84    0.07   22.84  63.45   1
239     2.91    0.02   23.49  63.45   1
240     2.86   -0.47   23.00  62.25   1
241     2.86    0.92   21.25  62.37   1
242     2.81    0.67   18.43  63.86   1
243     2.78   -0.30   18.88  65.62   1
244     2.81    0.01   19.46  64.96   1
245     2.87    0.24   17.86  63.66   1
246     2.87   -0.07   19.26  63.72   1
247     2.69   -0.07   20.99  63.81   1
248     2.60    0.05   20.64  64.45   1
249     2.71    0.22   20.71  63.69   1
250     2.75   -0.57   21.32  64.43   1

[251 rows x 5 columns]

First we select the zero, first, second, and fourth columns as the x-variable matrix
The third column as y

In [5]:
X = temp.iloc[:,[0,1,2,4]]
Y = temp.iloc[:,3]
Y = Y.values.reshape(len(df), 1)

In [6]:
beta_hat= dot(dot(inv(dot(X.T, X)), X.T), Y)

Then we get the beta value according to the normal equation

In [7]:
print(beta_hat)

[[  5.19791983]
 [ -0.16705303]
 [ -1.25623247]
 [ 74.38017535]]


Y = 5.19791983*X1 -0.16705303*X2 -1.25623247*X3 + 74.38017535

#### Defining the OLS function

In [8]:
def ols(Y, X):
    beta_hat = dot(dot(inv(dot(X.T, X)), X.T), Y)#inv(X.T * X) * X.T * Y
    y_hat = np.matmul(X, beta_hat) #X * beta_hat
    residuals = Y - y_hat
    print("residuals:")
    print(residuals)
    
    from numpy import sqrt
    sigma2 = (1 / 251) * np.dot(residuals.T,residuals)    #residuals.T * residuals
    sigma = sqrt(sigma2)
    print("sigma:")
    print(sigma)
    
    varcov_beta_hat2 = inv(dot(X.T,X))     #(sigma^2)*inv(X.T *X)
    print("varcov_beta_hat2")
    print(varcov_beta_hat2)#varcov_beta_hat = (sigma^2)*varcov_beta_hat2

    varcov_beta_hat = sigma*sigma*varcov_beta_hat2
    print("varcov_beta_hat")
    print(varcov_beta_hat)

    var_beta_hat = np.diag(varcov_beta_hat)
    print("var_beta_hat")
    print(var_beta_hat)
    
    a = np.dot(residuals.T,residuals)
    b = np.var(Y)
    print(a)
    print(b)
    R_square = 1 - ((np.dot(residuals.T,residuals)) / (np.var(Y)*251))
    print("R_square")
    print(R_square)
    
    adjust_R_square = 1-((1-R_square)*(251-1)/(251-4-1))
    print("adjust_R_square")
    print(adjust_R_square)

    print("beta_hat")
    print(beta_hat)
    
    print("var_beta_hat")
    print(var_beta_hat)
    
    print("np.std(var_beta_hat)")
    print(np.std(var_beta_hat))
    
    t_stat = beta_hat / (np.std(var_beta_hat))
    print("t_stat")
    print(t_stat)
    
    from scipy.stats import norm
    p_val_t = 1-norm.cdf(t_stat)
    print("p_val_t")
    print(p_val_t)
    
    F_stat = np.dot(dot(beta_hat.T, varcov_beta_hat),beta_hat)/(np.dot(residuals.T,residuals))
    print("F_stat")
    print(F_stat)
    
    from scipy.stats import chi2
    p_val_F = 1-chi2.cdf(F_stat,250)
    print("p_val_F")
    print(p_val_F)
    

In [9]:
ols(Y, X)

residuals:
[[  4.3884956 ]
 [  3.1965934 ]
 [  2.44455439]
 [  2.27670431]
 [  5.99926792]
 [  4.38320374]
 [  3.67727308]
 [  3.27894311]
 [  1.17728315]
 [  1.63022267]
 [  2.06762986]
 [  1.17358447]
 [ -1.36886155]
 [  2.50149592]
 [  1.30497074]
 [  2.2367519 ]
 [  0.16981944]
 [  1.01953698]
 [  1.96894652]
 [  3.01028603]
 [  3.78627243]
 [  1.95370837]
 [  0.74001375]
 [  2.12077886]
 [  2.6572716 ]
 [  0.21395325]
 [ -0.20498542]
 [ -3.30000173]
 [ -1.35813572]
 [ -0.85107603]
 [  0.71641277]
 [ -0.78768468]
 [ -3.38665473]
 [ -1.6959877 ]
 [  0.58182473]
 [  1.03300941]
 [ -0.24357628]
 [ -0.7755788 ]
 [ -1.44718814]
 [ -1.17627678]
 [ -1.75315862]
 [  0.29913192]
 [  0.12175948]
 [ -0.07520291]
 [  6.26047797]
 [  6.48039596]
 [  7.53498638]
 [  2.27079807]
 [  2.39424009]
 [ -2.15071338]
 [ -4.52198085]
 [ -4.97767623]
 [ -3.31133624]
 [ -0.4584588 ]
 [ -1.1635663 ]
 [ -2.88359399]
 [ -1.5956017 ]
 [ -0.78159773]
 [ -2.14124361]
 [ -3.65911546]
 [ -3.98551249]
 [ -4.1371471

As the result above:

        Y = 5.19791983*X1 -0.16705303*X2 -1.25623247*X3 + 74.38017535
        
t_stat:     0.90          -0.03          -0.22            12.95

p_val_t:    0.18          0.51           0.58             0

F_stat:     24.46230202

p_val_F:    1

R_square:   0.54300472

adj_R_squ:  0.5355739

The change of y can be 54% explain of x, so the simulation of this equation is pretty good.

Observed value is 251, according to t_table before the x1,x2,x3 values are less than the t_table， so we can not reject null hypothesis the coefficient is not different from zero, so the coefficiton for x is zero, and the coefficient for intercept is bigger than zero, so the coefficient is statistic significant.
